File: Unit Area of largest region of 1's.cpp
class Solution
{
    public:
    //Function to find unit area of the largest region of 1s.
    void solve(vector<vector<int>> &grid , int i , int j , int n , int m , int &cnt)
    {
        if(i < 0 || j < 0 || i >= n || j >= m || grid[i][j] == 0)
            return;
            
        cnt+=1;
        grid[i][j] = 0;
      
        // 8 directions
        solve(grid,i-1,j,n,m ,cnt);
        solve(grid,i+1,j,n,m ,cnt);
        solve(grid,i,j-1,n,m ,cnt);
        solve(grid,i,j+1,n,m ,cnt);
        solve(grid,i-1,j-1,n,m ,cnt);
        solve(grid,i-1,j+1,n,m ,cnt);
        solve(grid,i+1,j-1,n,m ,cnt);
        solve(grid,i+1,j+1,n,m ,cnt);
    }
  
    int findMaxArea(vector<vector<int>>& grid) 
    {
        int ans = 0;
        int n = grid.size() , m = grid[0].size();
        for(int i = 0;i<n;i++)
        {
            for(int j = 0;j<m;j++)
            {
                if(grid[i][j] == 1)
                {
                    int cnt = 0;
                    solve(grid , i , j , n , m , cnt);
                    ans = max(ans , cnt);
                }
            }
        }
        
        return ans;
        // Code here
    }
};

// class Solution
// {
//     public:
    
//     int n, m;
//     int dx[8] = {-1, -1, -1, 0, 0, 1, 1, 1};
//     int dy[8] = {-1, 0, 1, -1, 1, -1, 0, 1};
    
//     // Function to count the area of adjacent elements
//     int countArea(vector<vector<int>>& grid, int i, int j)
//     {
//         if(i < 0 or i >= n or j < 0 or j >= m or grid[i][j] == 0)
//         return 0;
        
//         int ans = 1;
//         grid[i][j] = 0;
        
//         for(int d=0; d<8; d++)
//             ans += countArea(grid, i+dx[d], j+dy[d]);
            
//         return ans;
//     }
    
//     //Function to find unit area of the largest region of 1s.
//     int findMaxArea(vector<vector<int>>& grid) {
//         // Code here
//         n = grid.size();
//         m = grid[0].size();
        
//         int ans = 0;
        
//         for(int i=0; i<n; i++)
//         {
//             for(int j=0; j<m; j++)
//             {
//                 if(grid[i][j] == 1)
//                 ans = max(ans, countArea(grid, i, j));
//             }
//         }
        
//         return ans;
//     }
// };

File: Unit Area of largest region of 1's.java
class Solution
{
    //Function to find unit area of the largest region of 1s.
    public int findMaxArea(int[][] grid)
    {
        // Code here
        int ans=0;
        boolean[][]vis=new boolean[grid.length][grid[0].length];
        for(int i=0;i<grid.length;i++){
            for(int j=0;j<grid[0].length;j++){
                if(!vis[i][j] && grid[i][j]==1){
                    ans=Math.max(ans,dfs(grid,i,j,vis));
                }
            }
        }
        return ans;
    }
    int dfs(int[][]grid,int i,int j,boolean[][]vis){
        if(!valid(grid,i,j,vis))return 0;
        vis[i][j]=true;
        int ans=1;
        ans+=dfs(grid,i+1,j,vis);
        ans+=dfs(grid,i-1,j,vis);
        ans+=dfs(grid,i,j+1,vis);
        ans+=dfs(grid,i,j-1,vis);
        ans+=dfs(grid,i+1,j+1,vis);
        ans+=dfs(grid,i+1,j-1,vis);
        ans+=dfs(grid,i-1,j+1,vis);
        ans+=dfs(grid,i-1,j-1,vis);
        return ans;
    }
    boolean valid(int[][]grid,int i,int j,boolean[][]vis){
        if(i<0 || j<0 || i==grid.length || j==grid[0].length
        || grid[i][j]==0 || vis[i][j])return false;
        return true;
    }
}

//Approach-2

class Solution
{
    int count;
    static boolean isSafe(int[][] grid,int row,int col,boolean[][] visited,int ROW,int COL)
    {
        if(row<0 || col<0 || row>=ROW || col>=COL || visited[row][col] || grid[row][col]==0)
            return false;
        return true;    
    }

    
    void DFS(int[][] grid,int row,int col,boolean[][] visited,int ROW,int COL)
    {
        int[] x = { -1, -1, -1, 0, 0, 1, 1, 1 };
        int[] y = { -1, 0, 1, -1, 1, -1, 0, 1 };

        visited[row][col] = true;

        for (int k = 0; k < 8; k++) 
        {
            if (isSafe(grid, row + x[k], col + y[k],visited, ROW, COL)) 
            {
                count++;
                DFS(grid, row + x[k], col + y[k],visited, ROW, COL);
            }
        }
    }
    
    public int findMaxArea(int[][] grid)
    {
        int ROW = grid.length;
        int COL = grid[0].length;
        
        boolean[][] visited = new boolean[ROW][COL];

        int result = 0;
      
        for (int i = 0; i < ROW; i++) 
        {
            for (int j = 0; j < COL; j++) 
            {
                
                if (grid[i][j] == 1 && !visited[i][j]) 
                {
                    count = 1;
                    DFS(grid, i, j, visited, ROW, COL);

                    result = Math.max(result, count);
                }
            }
        }
        return result;
    }
}

    
   
